4_visualize/src/plot_results.R

```diff
@@ -0,0 +1,412 @@
+#' @param input_wide The wide-formatted input data
+#' @return a munged dataframe of GLM predictions that includes all of 
+#' the columns in the original wide data except the 'temp_X' columns,
+#' which are instead reduced to two columns: 'depth' and 'temperature'
+munge_long <- function(input_wide) {
+  input_long <- input_wide %>%
+    pivot_longer(temp_0:colnames(input_wide)[ncol(input_wide)], names_to="depth", values_to="temperature") %>%
+    mutate(depth = as.numeric(str_remove(depth, 'temp_'))) %>%
+    arrange(date, depth)
+}
+
+#' @title Get the surface, middle, and bottom depths for a given lake, 
+#' for plotting. Modified from Alison's code for picking surface, middle, 
+#' and bottom depths from observational data: 
+#' https://code.usgs.gov/wma/wp/lake-temperature-neural-networks/-/blob/master/3_assess/src/inspect_site.R#L81-93
+#' @description For a given lake, identify which depths represent the
+#' surface, middle, and bottom sections of the lake
+#' @param predictions_long A long-format dataset of GLM predictions at
+#' 0.5-meter intervals
+#' @return A named list of the surface, middle, and bottom depths to be used
+#' to plot data for a given lake.
+get_plotting_depths <- function(predictions_long) {
+  n_lambda <- 1
+  depth_scores <- predictions_long %>% group_by(depth) %>% tally() %>%
+    mutate(
+      n_score = n_lambda * n/max(n),
+      surface_score = n_score - abs(depth - 0),
+      middle_score = n_score - abs(depth - max(depth)/3),
+      bottom_score = n_score - abs(depth - max(depth))
+    )
+  depths <- c(
+    surface = depth_scores$depth[which.max(depth_scores$surface_score)],
+    middle = depth_scores$depth[which.max(depth_scores$middle_score)],
+    bottom = depth_scores$depth[which.max(depth_scores$bottom_score)])
+  
+  plot_depth_surface = unname(depths['surface'])
+  plot_depth_middle = unname(depths['middle'])
+  plot_depth_bottom = unname(depths['bottom'])-0.5 # reduce by 0.5m to try to avoid NA values, which can occur at the bottom
+  return(list(plot_depth_surface=plot_depth_surface, plot_depth_middle=plot_depth_middle, plot_depth_bottom=plot_depth_bottom))
+}
+
+#' @title Create a violin chart of predicted surface, middle, and bottom 
+#' temperatures in all lakes, across GCMs
+#' @description Plot the distribution of predicted temperatures
+#' across all lakes, broken out by depth class (surface, middle or bottom),
+#' gcm, and time period. The output plot is faceted by GCM and by period
+#' @param run_groups The `p2_glm_uncalibrated_run_groups` grouped version 
+#' of the `p2_glm_uncalibrated_runs` output tibble subset to the site_id, 
+#' gcm, time_period, raw_meteo_fl, export_fl, and export_fl_hash columns 
+#' and grouped by site_id and gcm. Then filtered to only groups for which 
+#' glm_success==TRUE for all runs in that group.
+#' @param outfile The filepath for the exported png
+#' @return The filepath of the exported png 
+plot_temp_violin_gcms <- function(run_groups, outfile) {
```

Comment:
Seems like you may be able to combine this function + `plot_temp_violin_months()` by adding an argument specifying whether it is all GCMs or not, `all_gcms = TRUE`. Then, you could use an `if(all_gcms)` to decide to do the `glm_preds` munging or not and set the title. Might be more that I am missing, but more maintainable to group shared code (and documentation!) when possible.

Reply:
Hmm they both use data for all GCMs. It's just that they facet by different variables and for the months plot that requires filtering the data.

Reply:
Now combined into a single function